File: 5-Problem-A-Day/186. Reverse Words in a String II.java
class Solution {
    public void reverseWords(char[] s) {
        reverse(s, 0, s.length - 1);        
        int start = 0, end = -1;        
        for (int i = 0; i < s.length; i++) {
            if (s[i] == ' ') {
                reverse(s, start, i - 1);
                start = i + 1;
            }                
        }
        reverse(s, start, s.length - 1);
    }
    
    private void reverse(char[] s, int start, int end) {
        while (start < end) {
            char tmp = s[start];
            s[start] = s[end];
            s[end] = tmp;
            start++;
            end--;
        }
    }
}

File: 2021/186. Reverse Words in a String II.cpp
class Solution {
public:
    void reverseWords(vector<char>& s) {
        reverse(s.begin(), s.end());
        int start = 0;
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == ' ' || i + 1 == s.size()) {
                reverse(s.begin() + start, s.begin() + ((i + 1 == s.size()) ? i + 1 : i));
                start = i + 1;
            }
        }
    }
};

class Solution {
public:
    void reverseWords(vector<char>& s) {
        reverseWord(s, 0, s.size() - 1);
        int start = 0;
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == ' ' || i + 1 == s.size()) {
                reverseWord(s, start, (i + 1 == s.size()) ? i : i - 1);
                start = i + 1;
            }
        }
    }
    
    void reverseWord(vector<char>& s, int left, int right) {
        while (left < right) {
            swap(s[left], s[right]);
            left++;
            right--;
        }
    }
};